# Ames Housing Regression Project

## Test 1: Local Data Ingestion and Structural Verification
In this test, I intend to load the raw dataset from my local directory into a Pandas DataFrame. My objective is to verify that the file path is correct, ensure the data is loaded without corruption, and inspect the initial shape (total rows and columns) of the dataset.

In [46]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Modelos e Métricas para a Regressão
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, LinearRegression  # O modelo assertivo do seu projeto anterior!
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
print("[SUCCESS] Todas as bibliotecas foram carregadas com sucesso!")

[SUCCESS] Todas as bibliotecas foram carregadas com sucesso!


In [2]:
base_housing = pd.read_csv('../data/raw/train.csv')
base_housing

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1455,1456,60,RL,62.0,7917,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,8,2007,WD,Normal,175000
1456,1457,20,RL,85.0,13175,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,MnPrv,NaN,0,2,2010,WD,Normal,210000
1457,1458,70,RL,66.0,9042,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,GdPrv,Shed,2500,5,2010,WD,Normal,266500
1458,1459,20,RL,68.0,9717,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,4,2010,WD,Normal,142125


In [3]:
base_housing.info()

<class 'pandas.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   str    
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   str    
 6   Alley          91 non-null     str    
 7   LotShape       1460 non-null   str    
 8   LandContour    1460 non-null   str    
 9   Utilities      1460 non-null   str    
 10  LotConfig      1460 non-null   str    
 11  LandSlope      1460 non-null   str    
 12  Neighborhood   1460 non-null   str    
 13  Condition1     1460 non-null   str    
 14  Condition2     1460 non-null   str    
 15  BldgType       1460 non-null   str    
 16  HouseStyle     1460 non-null   str    
 17  OverallQual    1460 non-null   int64  
 18  OverallCond    1460

## 2. Exploratory Data Analysis (EDA)

### 2.1 Dataset Structure Evaluation
By running `base_housing.info()`, we verify that the dataset contains 1,460 records and 81 features (columns 0 to 80). 
* **Target Variable:** `SalePrice` (the home price we want to predict).
* **Feature Types:** A mix of continuous numerical variables (integers/floats) and categorical attributes (objects).
* **Missing Data:** Several columns show a `non-null` count lower than 1,460, indicating the presence of missing values (`NaN`) that require an imputation strategy.

In [4]:
base_housing.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [5]:
base_housing.isnull().sum()

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

### 2.2 Missing Values Identification
By analyzing `base_housing.isnull().sum()`, we can locate the exact volume of missing data per feature. 
* **Key Finding:** The feature `LotFrontage` shows a significant amount of missing data, with **259 null values** (`NaN`).
* **Next Step:** Since the standard output truncates the rows, we will filter and display only the columns that actually contain missing values to plan our imputation strategy.

In [6]:
# Filter and show only columns that contain missing values (NaN)
missing_counts= base_housing.isnull().sum()
missing_counts[missing_counts >0]

LotFrontage      259
Alley           1369
MasVnrType       872
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64

### 2.3 Comprehensive Missing Data Analysis
By filtering the features with missing values, we observe a clear distinction between random missingness and structural missingness:

* **Structural Missingness (Structural NaNs):** Features like `PoolQC` (1,453), `MiscFeature` (1,406), `Alley` (1,369), `Fence` (1,179), and `FireplaceQu` (690) have high null counts because most houses simply do not possess these physical attributes (e.g., no pool, no fence). The same pattern applies to the 81 missing rows in the `Garage` variables and 37/38 rows in the `Basement` variables.
* **Accidental/Random Missingness:** Features like `Electrical` (1 missing row), `MasVnrArea` (8 missing rows), and `LotFrontage` (259 missing rows) represent real missing information that must be handled via mathematical or statistical imputation.

## 3. Data Cleaning and Imputation Strategy
In this section, I will handle the missing values identified during the exploratory phase. My strategy is split into two parts:
1. **Categorical Imputation:** Replace `NaN` with a specific category (e.g., `'None'` or `'No'`) for houses that do not have the physical attribute.
2. **Numerical Imputation:** Replace `NaN` with statistical metrics (like the median) for continuous variables.

In [7]:
# Check the most frequent value(mode) in the Electrical column
most_frequent_electrical = base_housing['Electrical'].mode()[0]

In [8]:
# Fill the single NaN value with the most frequent value
base_housing['Electrical'] = base_housing['Electrical'].fillna(most_frequent_electrical)

In [9]:
# Verify if there are any missing values left in Electrical
base_housing['Electrical'].isnull().sum()

np.int64(0)

### 3.1 Imputation for Electrical Column
Since the `Electrical` feature had only 1 missing value, I identified the most frequent electrical system (the mode) in the dataset and used it to fill this single gap. After the imputation, I verified that the remaining null count for this column is 0.

In [10]:
# Inspect the unique values and counts for a garage column before imputation
base_housing['GarageType'].value_counts(dropna=False)

GarageType
Attchd     870
Detchd     387
BuiltIn     88
NaN         81
Basment     19
CarPort      9
2Types       6
Name: count, dtype: int64

In [11]:
# List of categorical garage columns to treat
garage_categorical_cols = ['GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']

# Fill NaN with 'None' for houses with no garage
for col in garage_categorical_cols:
    base_housing[col] = base_housing[col].fillna('None')

In [12]:
# Verify if NaN changed to 'None'
base_housing['GarageType'].value_counts(dropna=False)

GarageType
Attchd     870
Detchd     387
BuiltIn     88
None        81
Basment     19
CarPort      9
2Types       6
Name: count, dtype: int64

### 3.2 Imputation for Categorical Garage Features
I identified 4 categorical garage features (`GarageType`, `GarageFinish`, `GarageQual`, and `GarageCond`) with 81 missing values. Since these missing values represent the absence of a garage, I filled the `NaN` entries with the string `'None'`. A final check confirmed that all `NaN` values in these columns were successfully replaced.

In [13]:
# Check de summary statistics for the garage year built column
base_housing['GarageYrBlt'].describe()

count    1379.000000
mean     1978.506164
std        24.689725
min      1900.000000
25%      1961.000000
50%      1980.000000
75%      2002.000000
max      2010.000000
Name: GarageYrBlt, dtype: float64

In [14]:
# List of categorical basement columns to treat
basement_categorical_cols = ['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2']

# Fill NaN with 'None' for houses with no basement
for col in basement_categorical_cols:
    base_housing[col] = base_housing[col].fillna('None')

In [15]:
# Verify if NaN changed to 'None' in the basement quality column
base_housing['BsmtQual'].value_counts(dropna=False)

BsmtQual
TA      649
Gd      618
Ex      121
None     37
Fa       35
Name: count, dtype: int64

### 3.3 Imputation for Categorical Basement Features
I identified 5 categorical basement features (`BsmtQual`, `BsmtCond`, `BsmtExposure`, `BsmtFinType1`, and `BsmtFinType2`) containing between 37 and 38 missing values. Since these missing entries represent the structural absence of a basement, I filled the `NaN` values with the string `'None'`. A subsequent check on `BsmtQual` confirmed that the missing rows were successfully updated to `'None'`.

In [16]:
# Fill missing categorical values for Masonry Type with 'None'
base_housing['MasVnrType'] = base_housing['MasVnrType'].fillna('None')

In [17]:
# Fill missing numerical values for Masonry Veneer Area with 0
base_housing['MasVnrArea'] = base_housing['MasVnrArea'].fillna(0)

In [18]:
# Verify if there are any remainig null values in the Masonry Veneer columns
print(base_housing['MasVnrType'].isnull().sum())
print(base_housing['MasVnrArea'].isnull().sum())

0
0


### 3.4 Imputation for Masonry Veneer (MasVnr) Features
I addressed the missing values in the masonry veneer variables. For the categorical feature `MasVnrType`, I replaced the `NaN` entries with the string `'None'` to explicitly indicate the absence of masonry encasing. For the numerical feature `MasVnrArea`, I filled the missing entries with `0`, representing a zero square feet area for houses without this architectural feature. A final verification confirmed that both columns now contain 0 missing values.

In [19]:
# List of categorical columns representing additional or rare features
additional_features_cols = ['Alley', 'PoolQC', 'Fence', 'MiscFeature']

# Fill NaN with 'None' to indicate the house does not have the feature
for col in additional_features_cols:
    base_housing[col] = base_housing[col].fillna('None')

In [20]:
# Check the remaining missing values for these columns
base_housing[additional_features_cols].isnull().sum()

Alley          0
PoolQC         0
Fence          0
MiscFeature    0
dtype: int64

### 3.5 Imputation for Additional and Rare Features
I treated the features with the highest volume of missing data: `Alley`, `PoolQC`, `Fence`, and `MiscFeature`. Because these missing entries represent the absence of luxury or secondary property elements (like pools, alleys, fences, or sheds) rather than faulty data, I replaced all `NaN` values with the string `'None'`. A final check confirmed that the missing value count for all four columns successfully dropped to 0.

In [21]:
# Fill missing values fpr Fireplace Quality with 'None' for houses without a fireplace
base_housing['FireplaceQu'] = base_housing['FireplaceQu'].fillna('None')

In [22]:
# Check if there are any missing values left in the entire dataset
missing_values = base_housing.isnull().sum()
missing_values[missing_values > 0]

LotFrontage    259
GarageYrBlt     81
dtype: int64

In [23]:
# Fix the garage year built by replacing NaN with 0
base_housing['GarageYrBlt'] = base_housing['GarageYrBlt'].fillna(0)

In [24]:
# Check the median value for the lot frontage column
base_housing['LotFrontage'].median()

np.float64(69.0)

In [25]:
# Fill missing values in LotFrontage with the median value (69.0)
base_housing['LotFrontage'] = base_housing['LotFrontage'].fillna(base_housing['LotFrontage'].median())

In [26]:
# Check again if there are any missing values left in the entire dataset
missing_values = base_housing.isnull().sum()
missing_values[missing_values > 0]

Series([], dtype: int64)

### 3.6 Final Imputation for Remaining Numerical Features
To complete the data cleaning phase, I addressed the final two numerical columns with missing data. First, for `GarageYrBlt`, I imputed `0` for the 81 properties that do not have a garage, maintaining structural consistency with the categorical garage variables. Second, for `LotFrontage` (259 missing values), I applied a median imputation strategy ($69.0$ feet) to prevent outliers from distorting the dataset. A final validation across the entire dataframe confirmed that $100\%$ of missing values have been successfully resolved (`Series([], dtype: int64)`).

In [27]:
# Define the quality mapping dictionary (Ordinal Scale)
quality_mapping = {
    'None': 0,
    'Po': 1,  # Poor / Ruim
    'Fa': 2,  # Fair / Regular
    'TA': 3,  # Typical-Average / Médio
    'Gd': 4,  # Good / Bom
    'Ex': 5   # Excellent / Excelente
}

# List of columns that use exactly this scale
cols_to_map = ['ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond', 'HeatingQC', 'KitchenQual', 'FireplaceQu', 'GarageQual', 'GarageCond']

In [28]:
# Apply the mapping to all listed columns
for col in cols_to_map:
    base_housing[col] = base_housing[col].map(quality_mapping)

# Check if BsmtQual sucessfully tyrned into numbers
base_housing['BsmtQual'].value_counts()


BsmtQual
3    649
4    618
5    121
0     37
2     35
Name: count, dtype: int64

## 4. Feature Engineering: Categorical Encoding

### 4.1 Ordinal Encoding for Quality and Condition Features
To prepare the categorical data for machine learning algorithms, I implemented an ordinal encoding strategy for variables representing qualitative scales. I created a consistent mapping dictionary (`quality_mapping`) to translate qualitative ratings (`None`, `Po`, `Fa`, `TA`, `Gd`, `Ex`) into sequential integers from `0` to `5`. I applied this transformation across 9 distinct features simultaneously (`ExterQual`, `ExterCond`, `BsmtQual`, `BsmtCond`, `HeatingQC`, `KitchenQual`, `FireplaceQu`, `GarageQual`, and `GarageCond`). A value count verification on `BsmtQual` confirmed the successful conversion into numerical values.

In [29]:
# Apply One_hot Econding to All remaining categorical features automatically
base_housing_encoded = pd.get_dummies(base_housing, drop_first=True, dtype=int)

# Check the new shape of the dataset to see how many columns were created
base_housing_encoded.shape

(1460, 233)

In [30]:
# Display the first few rows of the fully numerical dataset
base_housing_encoded.head()

,Id,MSSubClass,LotFrontage,LotArea,OverallQual,OverallCond,YearBuilt,YearRemodAdd,MasVnrArea,ExterQual,...,SaleType_ConLI,SaleType_ConLw,SaleType_New,SaleType_Oth,SaleType_WD,SaleCondition_AdjLand,SaleCondition_Alloca,SaleCondition_Family,SaleCondition_Normal,SaleCondition_Partial
0,1,60,65.0,8450,7,5,2003,2003,196.0,4,...,0,0,0,0,1,0,0,0,1,0
1,2,20,80.0,9600,6,8,1976,1976,0.0,3,...,0,0,0,0,1,0,0,0,1,0
2,3,60,68.0,11250,7,5,2001,2002,162.0,4,...,0,0,0,0,1,0,0,0,1,0
3,4,70,60.0,9550,7,5,1915,1970,0.0,3,...,0,0,0,0,1,0,0,0,0,0
4,5,60,84.0,14260,8,5,2000,2000,350.0,4,...,0,0,0,0,1,0,0,0,1,0


### 4.2 One-Hot Encoding for Nominal Categorical Features
For the remaining nominal categorical variables (features with no intrinsic mathematical order, such as neighborhood, foundation type, or roof style), I applied a One-Hot Encoding strategy using the `pd.get_dummies()` function. I set the parameter `drop_first=True` to eliminate the first category of each variable, preventing the dummy variable trap and mitigating multicollinearity issues. To keep the dataset clean, I enforced an integer data type (`dtype=int`) for the newly generated binary flags. This automated expansion successfully transformed the dataframe from its original features into a fully numerical dataset with a shape of `(1460, 233)`, making it completely compatible with machine learning algorithms.

In [31]:
# Separete feature (X) and target variable(y)
X = base_housing_encoded.drop(columns=['SalePrice'])
y = base_housing_encoded['SalePrice']

print(f'Shape of X (feature): {X.shape}')
print(f'Shape of y (Target): {y.shape}')

Shape of X (feature): (1460, 232)
Shape of y (Target): (1460,)


In [32]:
# Split the dataset into 80% training and 20% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f'X_train shape: {X_train.shape} | y_train shape: {y_train.shape}')
print(f'X_test shape: {X_test.shape} | y_test shape: {y_test.shape}')

X_train shape: (1168, 232) | y_train shape: (1168,)
X_test shape: (292, 232) | y_test shape: (292,)


## 5. Dataset Splitting (Train-Test Split)

To evaluate the predictive performance of the machine learning models and prevent overfitting, I separated the dataset into features and the target variable. The target variable $y$ was defined as `SalePrice`, while the feature matrix $X$ retained the remaining 232 predictors. 

Using `sklearn.model_selection.train_test_split`, I divided the data into training ($80\%$) and testing ($20\%$) sets, enforcing data reproducibility with a `random_state` of 42. The resulting partitions yielded the following shapes:
* **X_train / y_train (Training Set):** `(1168, 232)` and `(1168,)` — used for model fitting.
* **X_test / y_test (Testing Set):** `(292, 232)` and `(292,)` — reserved for unbiased model evaluation.

In [33]:
# Instatiate the Linear Regression Model
lr_model = LinearRegression()

In [34]:
# train the Model using the training data
lr_model.fit(X_train, y_train)

print('Model trained successfully!')

Model trained successfully!


In [35]:
# Use the trained model to predict house prices for the test set
y_pred = lr_model.predict(X_test)

In [36]:
# Display the first 5 predictions vs the actual prices to see how it's doing
comparison_df = pd.DataFrame({'Actual Price': y_test, 'Predicted Price': y_pred})
comparison_df.head()

,Actual Price,Predicted Price
892,154500,155311.339505
1105,325000,330687.835137
413,115000,101139.355599
522,159000,177038.497862
1036,315500,304765.728598


## 6. Model Training and Baseline Evaluation

### 6.1 Linear Regression Implementation
With the training and testing datasets successfully partitioned, I initialized and trained a baseline Ordinary Least Squares (OLS) `LinearRegression` model using the training data ($X_{train}, y_{train}$). 

Following the training phase, I generated price predictions ($y_{pred}$) using the unseen test features ($X_{test}$). A preliminary, row-by-row comparative analysis between the `Actual Price` and the `Predicted Price` demonstrated that the model effectively captured the underlying pricing trends, yielding highly competitive estimates even before applying advanced feature scaling or regularization techniques.

### 6.2 Error Analysis and Business Context
I observed specific prediction variances (ranging from $14k to $18k) in properties such as houses 413 and 522. From a business perspective, these discrepancies often align with features that originally contained missing values (such as `FireplaceQu` or `BsmtQual`). By imputing `0` (representing 'None') during the encoding phase, I explicitly signaled the absence of these assets to the model. While this prevents data leakage and architectural errors, a simple linear model struggles to capture non-linear interactions (e.g., how the lack of a fireplace affects value differently across distinct neighborhoods), leading to these baseline deviations.

## 7. Next Steps: Non-Linear Modeling

### 7.1 Transition to Tree-Based Algorithms
To improve upon the baseline restrictions of the `LinearRegression` model—specifically its inability to capture complex, non-linear feature interactions (such as the trade-off between neighborhood quality and the presence of individual amenities like fireplaces)—the next phase of this project will implement tree-based models. I will train a `DecisionTreeRegressor` and subsequently ensemble models like `RandomForestRegressor` to map these conditional relationships and reduce the residual errors observed in the baseline.

In [38]:
# Istantiate the Decision Tree Model
# We set a max_depth to prevent the tree form growing infinitely and overfitting
dt_model = DecisionTreeRegressor(max_depth=5, random_state=42)

# Train the model
dt_model.fit(X_train, y_train)

print("Decision Tree Model trained successfully!")

Decision Tree Model trained successfully!


In [40]:
# Generate predictions with the Decision Tree
y_pred_dt = dt_model.predict(X_test)

In [41]:
# Add the new predictions to our comparison dataframe
comparison_df['Decision Tree Price'] = y_pred_dt
comparison_df.head()

,Actual Price,Predicted Price,Decision Tree Price
892,154500,155311.339505,147492.236364
1105,325000,330687.835137,434478.333333
413,115000,101139.355599,147492.236364
522,159000,177038.497862,187204.063380
1036,315500,304765.728598,344700.000000


### 7.2 Core Concept Analysis: Prediction Mechanics and Tree Depth

To build a rigorous foundation for this modeling phase, I analyzed two core mechanics of the Machine Learning pipeline:

* **The `.predict()` Function:** In the machine learning ecosystem, `predict` represents the inference phase. While `.fit()` extracts statistical patterns from the training data, `.predict()` applies those derived mathematical rules to unseen feature matrices ($X_{test}$) to estimate continuous numerical values (house prices). This acts as the final production engine of the data product.
* **Hyperparameter `max_depth=5`:** This parameter controls the structural complexity of the model by setting the maximum depth (number of split levels) of the decision tree. Restricting the depth to `5` prevents the algorithm from creating overly granular splits that memorize the training set. This acts as a vital regularization technique to mitigate **overfitting** and preserve the model's ability to generalize to new data.

### 7.3 Visualizing Tree Traversal and Decision Paths
To contextualize how the `DecisionTreeRegressor` operates, the `max_depth` parameter can be conceptualized as a maximum chain of consecutive branching decisions. When setting the depth to 5, the algorithm is constrained to traverse a maximum of 5 sequential branches (or binary questions) to isolate a property. Properties that share similar macroeconomic characteristics—such as houses 892 and 413—will follow the same conditional path and land within the exact same terminal leaf node, receiving an identical calculated mean price baseline (`147492.23`).

In [43]:
# list of the 5 house to analyzing
selected_houses = [892, 1105, 413, 522, 1036]

In [44]:
# Columns to inspect (using original feature form base_housing before ecoding for EASIER READING )
features_to_check = ['GrLivArea', 'OverallQual', 'BedroomAbvGr', 'Fireplaces', 'BsmtQual', 'SalePrice']

In [45]:
# Display the real data for these specific houses
base_housing.loc[selected_houses, features_to_check]

,GrLivArea,OverallQual,BedroomAbvGr,Fireplaces,BsmtQual,SalePrice
892,1068,6,3,0,3,154500
1105,2622,8,3,2,5,325000
413,1028,5,2,1,3,115000
522,1664,6,3,2,3,159000
1036,1620,9,2,1,5,315500


### 7.4 Empirical Validation of Leaf Node Assignment

By inspecting the underlying ground-truth features of the test instances, I validated the mathematical behavior of the tree slices:
* **Houses 892 and 413:** Both present highly similar dimensional profiles (`GrLivArea` of 1068 vs 1028 sqft) and baseline structures. This quantitative proximity causes them to follow the exact same structural splits up to `max_depth=5`, dropping into a single localized leaf that outputs their shared prediction.
* **House 1105:** This property exhibits extreme luxury metrics (`GrLivArea` of 2622 sqft, `OverallQual` of 8, and a maxed-out `BsmtQual` of 5). The tree isolation logic forced this sample into its premium terminal nodes. However, because decision trees create step-function boundaries rather than smooth continuous lines, it introduced a significant upward variance ($434k prediction vs $325k actual), highlighting the model's susceptibility to high variance when handling edge cases at shallow depths.

In [47]:
# Instantiate the Random Forest with 100 trees
rf_model = RandomForestRegressor(n_estimators=100, max_depth=5, random_state=43)

In [48]:
# Train the model
rf_model.fit(X_train, y_train)

print('Random Forest model trained successfully with 100 estimators!')

Random Forest model trained successfully with 100 estimators!


In [49]:
# Generate predictions with Random Forest
y_pred_rf = rf_model.predict(X_test)

In [50]:
# Add the new predictions to our comparison dataframe
comparison_df['Random Forest Price'] = y_pred_rf
comparison_df.head()

,Actual Price,Predicted Price,Decision Tree Price,Random Forest Price
892,154500,155311.339505,147492.236364,135154.922850
1105,325000,330687.835137,434478.333333,321706.506763
413,115000,101139.355599,147492.236364,129886.551475
522,159000,177038.497862,187204.063380,167033.166464
1036,315500,304765.728598,344700.000000,314605.184094


### 7.5 Advanced Evaluation: Ensemble Smoothness via RandomForestRegressor

By integrating a `RandomForestRegressor` with 100 estimators, I successfully addressed the high variance and step-function limitations inherent to the single decision tree model:
* **Mitigating Extreme Variance (House 1105):** Where the single tree overshot the real value by over $109k ($434,478 vs $325,000 actual), the Random Forest stabilized the prediction to **$321,706**. This demonstrates how bagging (bootstrap aggregating) averages out individual model anomalies.
* **Granular Differentiation (Houses 892 and 413):** The model broke the mathematical deadlock that previously assigned identical prices to these distinct rows. By utilizing randomized feature subsets across 100 parallel trees, the ensemble captured the subtle variations between the properties, resulting in distinct, highly optimized predictions closer to the historical truths.